Migrate build process to gradle

wallner opened the pull request:
This pull request swaps out the build process based on Maven and replaces it with a Gradle based one. This makes modifying the buildfile less painful and speeds up the build considerably.

wallner committed:
Migrate build process to gradle

tkrille commented:
@wallner Please rebase now.

tkrille commented:
Why no `gradlew.bat`? One should also be able to build OSIAM on Windows, shouldn't they?

tkrille commented:
If I run `gradle wrapper` or `./gradlew wrapper` I get a different `gradlew` script. It appears to be newer, as it contains an additional bug fix and supports an additional environment. The `gradle-wrapper.jar` also differs. Please run `./gradlew wrapper` again and check if this changes the aforementioned files too.

tkrille commented on build.gradle:
```diff
@@ -0,0 +1,187 @@
+
+group = 'org.osiam'
+
+description = """OSIAM"""
```
Description in POM was:

```
OSIAM is a secure identity management solution providing REST based services for authentication and
authorization. We achieve this by implementing two important open standards:

* OAuth 2.0
* SCIM 2.0

OSIAM is published under the MIT licence, giving you the greatest freedom possible to utilize OSIAM in you
project or product.
```

For what is the description used anyway? Maybe we can leave it off altogether?

wallner replied:
Let's drop the description. I agree.

tkrille commented on build.gradle:
```diff
@@ -0,0 +1,187 @@
+
+}
+war {
+    baseName = 'OSIAM'
```
This must be `osiam`.

tkrille commented on build.gradle:
```diff
@@ -0,0 +1,187 @@
+    compile 'org.springframework.boot:spring-boot-starter-thymeleaf'
+    compile 'org.springframework.boot:spring-boot-starter-security'
+    compile 'org.springframework.security.oauth:spring-security-oauth2'
+    compile 'org.springframework.boot:spring-boot-starter-tomcat'
```
This must be removed. It is defined again below with the right scope.

tkrille commented on build.gradle:
```diff
@@ -0,0 +1,187 @@
+    testCompile 'com.jayway.jsonpath:json-path:2.2.0'
+    testCompile 'nl.jqno.equalsverifier:equalsverifier:1.7.8'
+
+    providedRuntime("org.springframework.boot:spring-boot-starter-tomcat")
```
Shouldn't this be `providedRuntime 'org.springframework.boot:spring-boot-starter-tomcat'`, like the other entries?

tkrille commented:
I cannot find a replacement for `com.mycila:license-maven-plugin`.

tkrille commented:
I cannot find a replacement for `org.pitest:pitest-maven`.

tkrille commented:
I'm missing a `classes` JAR, i.e. a normal JAR file that contains alle classes of OSIAM. It was created before with this configuration in the POM: https://github.com/osiam/osiam/blob/master/pom.xml#L488

tkrille commented:
The WAR file previously contained `CHANGELOG.md` and `README.md`. I'm missing these files now.

wallner commented:
It was not my intention to build a one-to-one replacement for the maven build. I think, for instance, that dropping the fuzzing for now is a good thing, since we never really used it anyways. I could be convinced to add a license plugin, is that something we want?

tkrille commented:
Alright, fair enough. We can drop the Pitest plugin, as we never used it anyways.

Regarding the license plugin: If we want each file to begin with a license header, then we should automate it (which implies enforcing it). I'm not a big fan of a license header in each file, but it seems common practice. So I would say, that we add a similar plugin.

Regarding the `classes` JAR: This was once requested by @jvmlet to make it possible to embed OSIAM in a custom Spring app. We also offered the possibility to do that in https://github.com/osiam/osiam/issues/256. I'm not comfortable with just removing this artifact. At least, it must be listed in the changelog that "feature" was removed.

Regarding the `CHANGELOG.md` and `README.md` in the WAR: The WAR file is our only distributable artifact, so I think it must contain at least the changelog. Having the readme, too, is also a good starting point for users. If you think that this is too much clutter, then at least mention the removal in the changelog. If users need it, they will eventually redemand it.

tkrille commented on circle.yml:
```diff
@@ -4,37 +4,32 @@ machine:
 
 dependencies:
   override:
-    - mvn -U dependency:resolve dependency:resolve-plugins
```
Is there any replacement for this line? It would be nice to cache downloaded dependencies and build plugins.

wallner replied:
I don't think so.

tkrille replied:
Please check the [CircleCI docs](https://circleci.com/docs/language-java/#dependencies--tests): It is `./gradlew dependencies`.

tkrille commented on circle.yml:
```diff
@@ -4,37 +4,32 @@ machine:
+    - find . -type f -regex ".*/build/test-results/*.*xml" -exec cp {} $CIRCLE_TEST_REPORTS/junit/ \;
+    - cp -r build/reports/jacoco/ $CIRCLE_ARTIFACTS
+    - test -z $CODACY_PROJECT_TOKEN || codacy-coverage-reporter -l Java -r build/reports/jacoco/test/jacocoTestReport.xml --projectToken $CODACY_PROJECT_TOKEN
+    - test -z $COVERALLS_REPO_TOKEN || ./gradlew jacocoTestReport -DrepoToken=$COVERALLS_REPO_TOKEN
```
I think the task must be `coveralls`, not `jacocoTestReport`. And is `-DrepoToken=$COVERALLS_REPO_TOKEN` really needed? It runs perfectly fine on my machine with just the environment variable defined.

wallner replied:
Probably not, but I tried to change as little here as possible, so I left the Property untouched. Removed now.

tkrille commented on circle.yml:
```diff
@@ -4,37 +4,32 @@ machine:
-    - cp -r target/coverage-reports/jacoco/ $CIRCLE_ARTIFACTS
-    - test -z $CODACY_PROJECT_TOKEN || codacy-coverage-reporter -l Java -r target/coverage-reports/jacoco/jacoco.xml --projectToken $CODACY_PROJECT_TOKEN
-    - test -z $COVERALLS_REPO_TOKEN || mvn coveralls:report -DrepoToken=$COVERALLS_REPO_TOKEN
+    - find . -type f -regex ".*/build/test-results/*.*xml" -exec cp {} $CIRCLE_TEST_REPORTS/junit/ \;
```
The Circle CI docs state that it must be

```
- find . -type f -regex ".*/build/test-results/.*xml" -exec cp {} $CIRCLE_TEST_REPORTS/junit/ \;
```

I.e. there's a `*` too much before `.*xml`.

wallner replied:
Well, then it was wrong in the first place.

tkrille replied:
No, it wasn't. The previous version was

```
find . -type f -regex ".*/target/.*-reports/.*xml" -exec cp {} $CIRCLE_TEST_REPORTS/junit/ \;
```

which contains a correct regex. Nevertheless, please remove the superfluous `*`.

tkrille commented on circle.yml:
```diff
@@ -4,37 +4,32 @@ machine:
-      - wget https://raw.githubusercontent.com/osiam/circleci/master/settings.xml
-      - mvn deploy:deploy-file -DrepositoryId=releases -Durl="https://api.bintray.com/maven/osiam/OSIAM/org.osiam:osiam/;publish=1" -s settings.xml
-      - curl -T target/osiam.war -u${BINTRAY_USER}:${BINTRAY_KEY} -H "X-Bintray-Package:osiam" -H "X-Bintray-Version:${CIRCLE_TAG}" -H "X-Bintray-Publish:1" -H "X-Bintray-Override:1" https://api.bintray.com/content/osiam/downloads/osiam/${CIRCLE_TAG}/osiam-${CIRCLE_TAG}.war
+      - ./gradlew -Pversion=${CIRCLE_TAG} bintrayUpload
```
Using this task doesn't automatically publish the uploaded artifacts. Please make sure that artifacts are published.

tkrille commented on src/main/antlr/org/osiam/storage/parser/LogicalOperatorRules.g4:
```diff
@@ -24,6 +24,10 @@
 
+@header {
+package org.osiam.storage.parser;
+}
```
Why is this suddenly needed?

wallner replied:
The Gradle antlr plugin does a little less magic than the maven one. Which I consider a good thing.

tkrille replied:
Alright, thanks for clarification.

tkrille commented on build.gradle:
```diff
@@ -0,0 +1,185 @@
+    compile 'org.springframework.boot:spring-boot-starter-security'
+    compile 'org.springframework.boot:spring-boot-starter-actuator'
+    compile 'org.springframework.security.oauth:spring-security-oauth2'
+    compile 'org.springframework.boot:spring-boot-starter-tomcat'
```
This must be removed. It is defined again below with the right scope.

wallner commented:
Ok, I added the `gradlew.bat` and updated the wrapper. I'm ok with adding a license plugin, however I don't really see a value in adding the `CHANGELOG.md` and `README.md` to the war-file. Do you really believe people are extracting that? Somehow I doubt it.

tkrille commented on build.gradle:
```diff
@@ -0,0 +1,185 @@
+    }
+
+}
+war {
```
Is there an easy way to just name the resulting WAR file `osiam.war`. This was the previous behavior and I found it kinda convenient.

tkrille commented:
> however I don't really see a value in adding the `CHANGELOG.md` and `README.md` to the war-file. Do you really believe people are extracting that? Somehow I doubt it.

Yeah, I'm not so sure anymore either. When in doubt, we don't do it. Let's leave it off.

dacrome commented:
hm, but the war is our released artifact. If it's no big deal, please add it.

wallner commented:
I don't think it make sense to include `README.md` and `CHANGELOG.md`. I can't see a reason for it. Do you really think people are unpacking the war file? I don't. I know at least I wouldn't. If you insist on providing an artefact that contains both files, I suggest that we build one archive that contains the `war` and the two files separately.

dacrome commented:
Yes, you are right. Leave it off.

wallner commented:
I found a license-plugin replacement and added it.

tkrille commented on build.gradle:
```diff
@@ -0,0 +1,195 @@
+license {
+    mapping {
+        g4 = 'JAVADOC_STYLE'
+        excludes (["**/*_.java", "**/LogicalOperatorRules*.java", "**/*.json"])
```
This line must be outside the closure for `mapping`. Though it works, it does so merely by accident.

wallner replied:
That's fixed now.

tkrille commented on circle.yml:
```diff
@@ -4,37 +4,33 @@ machine:
-      - mvn deploy:deploy-file -DrepositoryId=snapshots -Durl=https://oss.jfrog.org/artifactory/oss-snapshot-local -s settings.xml
-      - curl -T target/osiam.war -u${BINTRAY_USER}:${BINTRAY_KEY} -H "X-Bintray-Package:osiam" -H "X-Bintray-Version:latest" -H "X-Bintray-Publish:1" -H "X-Bintray-Override:1" https://api.bintray.com/content/osiam/downloads/osiam/latest/osiam-latest.war
+      - ./gradlew publish
+      - curl -T build/libs/OSIAM.war -u${BINTRAY_USER}:${BINTRAY_KEY} -H "X-GPG-PASSPHRASE:${GPG_PASSPHRASE}" -H "X-Bintray-Package:osiam" -H "X-Bintray-Version:latest" -H "X-Bintray-Publish:1" -H "X-Bintray-Override:1" https://api.bintray.com/content/osiam/downloads/osiam/latest/osiam.war
```
This must be `curl -T build/libs/osiam.war -u${BINTRAY_USER} …` now. The WAR file is lower-cased again.

tkrille replied:
The URL to POST to must be `https://api.bintray.com/content/osiam/downloads/osiam/latest/osiam-latest.war` like before.

tkrille commented:
Alright, looks pretty good to me now. There are 3 little things remaining. Please fix them and I'll merge this immediately.

dacrome commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,6 @@
+distributionPath=wrapper/dists
+zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-3.1-all.zip
```
Can you use Gradle 3.2?

tkrille commented on circle.yml:
```diff
@@ -4,37 +4,33 @@ machine:
-      - wget https://raw.githubusercontent.com/osiam/circleci/master/settings.xml
-      - mvn deploy:deploy-file -DrepositoryId=snapshots -Durl=https://oss.jfrog.org/artifactory/oss-snapshot-local -s settings.xml
-      - curl -T target/osiam.war -u${BINTRAY_USER}:${BINTRAY_KEY} -H "X-Bintray-Package:osiam" -H "X-Bintray-Version:latest" -H "X-Bintray-Publish:1" -H "X-Bintray-Override:1" https://api.bintray.com/content/osiam/downloads/osiam/latest/osiam-latest.war
+      - ./gradlew publish
```
This will upload a WAR that has not been enhanced and repackaged by Spring Boot. Either make this `./gradlew assemble publish` or let the `publish` task depend on `assemble`.

tkrille commented on circle.yml:
```diff
@@ -4,37 +4,33 @@ machine:
-      - wget https://raw.githubusercontent.com/osiam/circleci/master/settings.xml
-      - mvn deploy:deploy-file -DrepositoryId=releases -Durl="https://api.bintray.com/maven/osiam/OSIAM/org.osiam:osiam/;publish=1" -s settings.xml
-      - curl -T target/osiam.war -u${BINTRAY_USER}:${BINTRAY_KEY} -H "X-Bintray-Package:osiam" -H "X-Bintray-Version:${CIRCLE_TAG}" -H "X-Bintray-Publish:1" -H "X-Bintray-Override:1" https://api.bintray.com/content/osiam/downloads/osiam/${CIRCLE_TAG}/osiam-${CIRCLE_TAG}.war
+      - ./gradlew -Pversion=${CIRCLE_TAG} bintrayUpload
```
See https://github.com/osiam/osiam/pull/270/files#r88811799

tkrille commented on circle.yml:
```diff
@@ -4,37 +4,35 @@ machine:
-      - mvn deploy:deploy-file -DrepositoryId=snapshots -Durl=https://oss.jfrog.org/artifactory/oss-snapshot-local -s settings.xml
-      - curl -T target/osiam.war -u${BINTRAY_USER}:${BINTRAY_KEY} -H "X-Bintray-Package:osiam" -H "X-Bintray-Version:latest" -H "X-Bintray-Publish:1" -H "X-Bintray-Override:1" https://api.bintray.com/content/osiam/downloads/osiam/latest/osiam-latest.war
+      - ./gradlew assemble
+      - ./gradlew publish
```
Unfortunately, `publish` depends on `war`, because of the `from components.web` statement in the publication's definition. Because the output of the original `war` task has changed (Spring Boot repackaged the WAR file), the `war` task will run again, but this time without the repackaging. This produces a broken WAR file that gets published. I'd suggest to merge the 2 lines: `./gradlew assemble publish`. This way, the repackaged WAR file is published. Please note that this is not true for the `bintrayUpload` task. So the next line can stay as it is.

wallner replied:
Thank you for looking into this.

tkrille commented on circle.yml:
```diff
@@ -4,37 +4,35 @@ machine:
-      - mvn deploy:deploy-file -DrepositoryId=releases -Durl="https://api.bintray.com/maven/osiam/OSIAM/org.osiam:osiam/;publish=1" -s settings.xml
-      - curl -T target/osiam.war -u${BINTRAY_USER}:${BINTRAY_KEY} -H "X-Bintray-Package:osiam" -H "X-Bintray-Version:${CIRCLE_TAG}" -H "X-Bintray-Publish:1" -H "X-Bintray-Override:1" https://api.bintray.com/content/osiam/downloads/osiam/${CIRCLE_TAG}/osiam-${CIRCLE_TAG}.war
+      - ./gradlew -Pversion=${CIRCLE_TAG} assemble
+      - ./gradlew -Pversion=${CIRCLE_TAG} publish
```
See https://github.com/osiam/osiam/pull/270/files#r89203451.

tkrille commented:
Besides my last two comments, everything works as expected.

tkrille commented:
Thank you very much!
